caches/infinispan/src/main/java/org/commonjava/maven/galley/cache/infinispan/FastLocalCacheProvider.java

```diff
@@ -316,71 +316,113 @@ public boolean isFile( ConcreteResource resource )
     public InputStream openInputStream( final ConcreteResource resource )
             throws IOException
     {
-        //use "this" as lock is heavy, should think about use the transfer for the resource as the lock for each thread
+        final String pathKey = getKeyForResource( resource );
+
+        // This lock is used to control the the local resource can be opened successfully finally when local resource missing
+        // but NFS not, which means will do a NFS->local copy.
+        final Object copyLock = new Object();
+        // A flag to mark if the local resource can be open now or need to wait for the copy thread completes its work
+        final AtomicBoolean canStreamOpen = new AtomicBoolean( false );
+        // This copy task is responsible for the NFS->local copy, and will be run in another thread,
+        // which can use PartyLine concurrent read/write function on the local cache to boost
+        // the i/o operation
+        final Runnable copyNFSTask = new Runnable()
+        {
+            @Override
+            public void run()
+            {
+                InputStream nfsIn = null;
+                OutputStream localOut = null;
+
+                try
+                {
+                    cacheTxMgr.begin();
+                    nfsOwnerCache.getAdvancedCache().lock( pathKey );
+                    File nfsFile = getNFSDetachedFile( resource );
+                    if ( !nfsFile.exists() )
+                    {
+                        logger.debug( "NFS cache does not exist too." );
+                        return;
+                    }
+                    nfsIn = new FileInputStream( nfsFile );
+                    localOut = plCacheProvider.openOutputStream( resource );
+                    IOUtils.copy( nfsIn, localOut );
+                    logger.debug( "NFS copy to local cache done." );
+                }
+                catch ( NotSupportedException | SystemException | IOException e )
+                {
+                    if ( e instanceof IOException )
+                    {
+                        final String errorMsg =
+                                String.format( "[galley] got i/o error when doing the NFS->Local copy for resource %s",
+                                               resource.toString() );
```

Comment:
Just use resource here, forget the toString() call. If resource is null, it'll render "null". Otherwise, it will only call toString() on the object if the log call proceeds (if the log level isn't disabled). Calling it preemptively like this incurs the string rendering cost no matter what.